[IsDeprecatedWeakRefSmartPointerException] Make FrameLoadState and subclasses RefCounted or equivalent

basuke opened the pull request:
#### 9af1550239785975e253bb8fc965060f7f05aa9e
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make FrameLoadState and subclasses RefCounted or equivalent
<a href="https://bugs.webkit.org/show_bug.cgi?id=281095">https://bugs.webkit.org/show_bug.cgi?id=281095</a>
<a href="https://rdar.apple.com/problem/137548780">rdar://problem/137548780</a>

Reviewed by Ryosuke Niwa and Chris Dumez.

Removing IsDeprecatedWeakRefSmartPointerException for FrameLoadState and make those classes ref/defef.

* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
(WebKit::FrameLoadState::didSuspend):
(WebKit::FrameLoadState::didReceiveServerRedirectForProvisionalLoad):
(WebKit::FrameLoadState::didFailProvisionalLoad):
(WebKit::FrameLoadState::didCommitLoad):
(WebKit::FrameLoadState::didFinishLoad):
(WebKit::FrameLoadState::didFailLoad):
(WebKit::FrameLoadState::setURL):
(WebKit::FrameLoadState::forEachObserver):
* Source/WebKit/UIProcess/FrameLoadState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxyFrameLoadStateObserver::WebPageProxyFrameLoadStateObserver):
(WebKit::WebPageProxyFrameLoadStateObserver::ref const):
(WebKit::WebPageProxyFrameLoadStateObserver::deref const):
(WebKit::PageLoadTimingFrameLoadStateObserver::ref const):
(WebKit::PageLoadTimingFrameLoadStateObserver::deref const):
(WebKit::WebPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: <a href="https://commits.webkit.org/284995@main">https://commits.webkit.org/284995@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04fcc171b4186044cd0c33b993f89cf4028a8770

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71097 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50511 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23871 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75203 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22302 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58311 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22122 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56225 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14701 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74163 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45910 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61292 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42578 "Passed tests") | | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20643 "Hash 04fcc171 for PR 34871 does not build (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64477 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19108 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76925 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15332 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18292 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63953 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15374 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/61339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63914 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15752 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12046 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5684 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46311 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1090 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47382 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48665 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47124 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
```diff
@@ -43,6 +43,9 @@ class SubFrameSOAuthorizationSession final : public NavigationSOAuthorizationSes
 
     ~SubFrameSOAuthorizationSession();
 
+    void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ref(); } \
```
What's up with the trailing `\`?

rniwa commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -140,13 +143,22 @@ class WebPageProxyFrameLoadStateObserver final : public FrameLoadStateObserver {
             m_visitedDomains.removeLast();
     }
 
+    const WebPageProxy& m_page;
```
Can we use WeakRef for m_page?

rniwa commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -182,6 +194,7 @@ class PageLoadTimingFrameLoadStateObserver final : public FrameLoadStateObserver
         // FIXME: Assert that m_loadingFrameCount is zero if this is a main frame.
     }
 
+    const WebPageProxy& m_page;
```
Ditto.

rniwa reviewed (approved):
r=me assuming we can use WeakRef for m_page.

cdumez commented on Source/WebKit/UIProcess/FrameLoadState.cpp:
```diff
@@ -51,9 +51,9 @@ void FrameLoadState::didStartProvisionalLoad(const URL& url)
-    m_observers.forEach([&url](FrameLoadStateObserver& observer) {
-        observer.didReceiveProvisionalURL(url);
-        observer.didStartProvisionalLoad(url);
+    forEachObserver([&url](Ref<FrameLoadStateObserver> observer) {
```
We shouldn't change the parameter type of `forEachObserver()`, there is no reason to do so. The call site needs a smart pointer, that's it.

basuke replied:
By defining `forEachObserver`, protected observer is on the stack when callback is called. This is similar pattern with protectedXXXX(). With this member function, most of the house keeping task is hidden from the callers. I think it's worth defining this.

I think better way is defining the similar function in WeakHashSet, like `forEachProtected(Ref<T>)`.

basuke replied:
Maybe simple for loop is simpler and clear.

cdumez commented on Source/WebKit/UIProcess/FrameLoadState.cpp:
```diff
@@ -51,20 +51,19 @@ void FrameLoadState::didStartProvisionalLoad(const URL& url)
-        observer.didReceiveProvisionalURL(url);
-        observer.didStartProvisionalLoad(url);
-    });
+    for (Ref observer : m_observers) {
```
This doesn't look safe to me since we're no longer copying the observers before iteration. As a result, if an observer unregisters itself during the loop, we'll crash.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -497,13 +497,47 @@ StorageRequests& StorageRequests::singleton()
+
+void WebPageProxyFrameLoadStateObserver::ref() const
+{
+    protectedPage()->ref();
```
Can't we just call `m_page->ref()`?

basuke replied:
Okay. Try static analysis before landing.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -497,13 +497,47 @@ StorageRequests& StorageRequests::singleton()
+
+void WebPageProxyFrameLoadStateObserver::deref() const
+{
+    protectedPage()->deref();
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -497,13 +497,47 @@ StorageRequests& StorageRequests::singleton()
 
+void PageLoadTimingFrameLoadStateObserver::ref() const
+{
+    protectedPage()->ref();
```
ditto

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -497,13 +497,47 @@ StorageRequests& StorageRequests::singleton()
+
+void PageLoadTimingFrameLoadStateObserver::deref() const
+{
+    protectedPage()->deref();
```
ditto.

cdumez reviewed (approved):
LGTM but please see my comments.

basuke committed:
[IsDeprecatedWeakRefSmartPointerException] Make FrameLoadState and subclasses RefCounted or equivalent
https://bugs.webkit.org/show_bug.cgi?id=281095
rdar://problem/137548780

Reviewed by Ryosuke Niwa and Chris Dumez.

Removing IsDeprecatedWeakRefSmartPointerException for FrameLoadState and make those classes ref/defef.

* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.h:
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
(WebKit::FrameLoadState::didSuspend):
(WebKit::FrameLoadState::didReceiveServerRedirectForProvisionalLoad):
(WebKit::FrameLoadState::didFailProvisionalLoad):
(WebKit::FrameLoadState::didCommitLoad):
(WebKit::FrameLoadState::didFinishLoad):
(WebKit::FrameLoadState::didFailLoad):
(WebKit::FrameLoadState::setURL):
(WebKit::FrameLoadState::forEachObserver):
* Source/WebKit/UIProcess/FrameLoadState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxyFrameLoadStateObserver::WebPageProxyFrameLoadStateObserver):
(WebKit::WebPageProxyFrameLoadStateObserver::ref const):
(WebKit::WebPageProxyFrameLoadStateObserver::deref const):
(WebKit::PageLoadTimingFrameLoadStateObserver::ref const):
(WebKit::PageLoadTimingFrameLoadStateObserver::deref const):
(WebKit::WebPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: https://commits.webkit.org/284995@main
